internal/expr/parser.go

```diff
@@ -532,18 +539,43 @@ func (p *Parser) parseOutputExpression() (*outputPart, bool, error) {
 	return nil, false, nil
 }
 
-// parseInputExpression parses an input expression of the form "$Type.name".
+// parseInputExpression parses an input expression of the form "Type.name".
 func (p *Parser) parseInputExpression() (*inputPart, bool, error) {
 	cp := p.save()
 
+	if fn, ok, err := p.parseSourceType(); ok {
+		return &inputPart{sourceType: fn, raw: p.input[cp.pos:p.pos]}, true, nil
+	} else if err != nil {
+		return nil, false, err
+	}
+
+	cp.restore()
+	return nil, false, nil
+}
+
+func (p *Parser) parseSourceType() (fullName, bool, error) {
 	if p.skipByte('$') {
 		if fn, ok, err := p.parseGoFullName(); ok {
-			if fn.name == "*" {
-				return nil, false, fmt.Errorf(`asterisk not allowed in input expression "$%s"`, fn)
```

Comment:
It's not clear that this is worth losing.  It doesn't seem to make much sense to have people saying "foo IN $Type.*", does it?  Try to picture what kind of type a developer would create that would hold homogeneous fields that were supposed to  indiscriminately be injected into SQL expressions as more fields are added?

Reply:
$Type.* is still disallowed (see [test](https://github.com/canonical/sqlair/pull/76/files#diff-060008e27564bf048cc01912a18a63f283931138073eb81ef7ad5f9c2141df6cR545-R547)), the difference now is that the logic has been moved from the parser itself to the prepare. Before we knew that a star was directly an error but now we need the type to see if it is a slice (valid) or a map/struct (not valid).